### PR TITLE
SRDF Volume add and Remove + Secure Snap

### PR DESCRIPTION
**Pull request by rawstorage**
added provisioning changes for adding and removing volumes from srdf groups CI_Tests and Unit_tests added change log updated and version incremented to 9.1.2

Secure Snap has also been added but no CI_TEST as this feature can't support cleanup.

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -1,7 +1,7 @@
 PyU4V Change Log
 ================
 
-Version 9.1.1.0 - released 12/12/19
+Version 9.1.2.0 - released 12/12/19
 ===================================
```
Comment (helenwalsh) [inline ChangeLog:26]: Date will need changing

File: PyU4V/provisioning.py [hunk 2]
```diff
@@ -1792,6 +1794,16 @@ def add_new_volume_to_storage_group(
         :param _async: if call should be async -- bool
         :param vol_name: name to give to the volume, optional -- str
         :param create_new_volumes: new volumes only, no ro-use -- bool
+        :param remote_array_1_id 12 digit serial number of remote array,
+        optional -- str
```
Comment (helenwalsh) [inline PyU4V/provisioning.py]: warning in generation of docstrings here (and below).  You need to align 'optional' with remote_array_1_id just above it

/currentdev/pyu4v_external/PyU4V/PyU4V/provisioning.py:docstring of PyU4V.provisioning.ProvisioningFunctions.add_new_volume_to_storage_group:10: WARNING: Field list ends without a blank line; unexpected unindent.

File: PyU4V/provisioning.py [hunk 3]
```diff
@@ -1792,6 +1794,16 @@ def add_new_volume_to_storage_group(
...
+        :param remote_array_1_sgs list of storage groups on remote array to add
+        Remote device, Unisphere instance must be local to R1 storage group
+        otherwise volumes will only be added to the local group.
+        optional -- str or list
```
Comment (helenwalsh) [inline PyU4V/provisioning.py]: and here

File: PyU4V/provisioning.py [hunk 4]
```diff
@@ -1792,6 +1794,16 @@ def add_new_volume_to_storage_group(
...
+        :param remote_array2_id optional digit serial number of remote array,
+        only used in multihop SRDF, e.g. R11, or R1 - R21 - R2 optional -- str
```
Comment (helenwalsh) [inline PyU4V/provisioning.py]: and here

File: PyU4V/provisioning.py [hunk 5]
```diff
@@ -1792,6 +1794,16 @@ def add_new_volume_to_storage_group(
...
+        :param remote_array2_sgs storage groups on remote array, optional
+        -- str or list
```
Comment (helenwalsh) [inline PyU4V/provisioning.py]: and here

File: PyU4V/provisioning.py [hunk 6]
```diff
@@ -1841,19 +1864,55 @@ def remove_vol_from_storagegroup(self, sg_id, vol_id, _async=False):
...
+            self, storage_group_id, vol_id, _async=False,
+            remote_array_1_id=None, remote_array_1_sgs=None,
+            remote_array_2_id=None, remote_array_2_sgs=None):
         """Remove a volume from a given storage group.
 
         :param storage_group_id: storage group id -- str
         :param vol_id: device id -- str
         :param _async: if call should be async -- bool
+        :param remote_array_1_id 12 digit serial number of remote array,
+        optional -- str
```
Comment (helenwalsh) [inline PyU4V/provisioning.py]: see comment above

File: PyU4V/replication.py [hunk 7]
```diff
@@ -226,11 +227,16 @@ def create_storage_group_snapshot(
         :param snap_name: snapshot name -- str
         :param ttl: Time To Live -- str
         :param hours: if TTL is in hours instead of days -- bool
+        :param secure: sets secure snapshot, snapshot created with secure
+        option can not be deleted before ttl expires -- bool
```
Comment (helenwalsh) [inline PyU4V/replication.py]: See comment above

**Review by helenwalsh — CHANGES_REQUESTED** [review]
besides CI issues just a few minor nits.

File: PyU4V/provisioning.py (hunk 2 above)
Comment (rawstorage) [reply PyU4V/provisioning.py]: will do..  see that in a few places

File: ChangeLog (hunk 1 above)
Comment (rawstorage) [reply ChangeLog:26]: I'll update the date just before the push to master

Comment (MichaelMcAleer) [reply ChangeLog:26]: The change shouldn't be made to existing change log entries but instead put at the top of the file so there is a full record of versions and the changes they include, we do something similar for our Cinder driver: https://github.com/openstack/cinder/blob/master/cinder/volume/drivers/dell_emc/powermax/fc.py#L37

File: PyU4V/replication.py (hunk 7 above)
Comment (MichaelMcAleer) [reply PyU4V/replication.py]: Still needs addressed

**Review by MichaelMcAleer — CHANGES_REQUESTED** [review]
Two changes need addressed, also the docs will need to be rebuilt to reflect changes made in the .rst files. Whats your coverage like for the unit tests? Are we still sitting at that nice 100%?

File: ChangeLog (hunk 1 above)
Comment (MichaelMcAleer) [reply ChangeLog:26]: Still needs changed, its an apache license requirement to have a version specific change log, it should be:

```
PyU4V Change Log
================

Version 9.1.2.0 - released 19/12/19
======================

[DETAILS]


Version 9.1.1.0 - released 12/12/19
======================

[DETAILS]

Version 9.1.0.0 - released 4/12/19
======================

[DETAILS]
```

File: PyU4V/replication.py (hunk 7 above)
Comment (MichaelMcAleer) [reply PyU4V/replication.py]: This is still not correct, you should be seeing sphinx build warnings at the very least for things like this I believe, did you try building them yet to make sure its building cleanly?

**Review by MichaelMcAleer — APPROVED** [review]
Looks good now Paul

**Review by odonos12 — APPROVED** [review]
All requests appear to be resolved now.

**Pull request merged**